numpy/core/src/multiarray/ctors.c

```diff
@@ -1259,6 +1259,52 @@ PyArray_New(PyTypeObject *subtype, int nd, npy_intp *dims, int type_num,
     return new;
 }
 
+int
+_copy_from_pil_style(char* dest, int ndim, char *buf, npy_intp *shape,
+                     npy_intp *strides, npy_intp *suboffsets, npy_intp itemsize)
+{
+    int ax, last_axis;
+    npy_intp indices[NPY_MAXDIMS];
+    void **pointers[NPY_MAXDIMS];
+    char *pointer = (char*)buf;
+    npy_intp blocksize;
+    /* Initialize indices and pointers. */
+    for (ax = 0; ax < ndim; ++ax) {
+        indices[ax] = 0;
+        pointers[ax] = (void **)pointer;
+        if (suboffsets[ax] >= 0) {
+            last_axis = ax;
+	        pointer = *((char**)pointer) + suboffsets[ax];
+	    }
+    }
+    /* compute blocksize */
+    blocksize = itemsize;
+    for (ax = last_axis + 1; ax < ndim; ++ax)
```

Comment:
suboffsets cannot be all negative because this would mean no dereferencing and the PEP requires that suboffsets pointer be NULL in this case.  If not for that requirement, the right initial value would indeed be `-1`, but I don't want to introduce confusion between real `-1` and `-1` that means `ndim - 1`, so I just initialize `last_axis` to `0`.